SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityMesosStatusUpdateHandler.java

```diff
@@ -254,21 +254,30 @@ private void unsafeProcessStatusUpdate(Protos.TaskStatus status) {
       scheduler.handleCompletedTask(task, taskIdObj, isActiveTask, timestamp, taskState, taskHistoryUpdateCreateResult, status);
     }
 
-    System.out.println(newTaskStatusHolder);
-
     saveNewTaskStatusHolder(taskIdObj, newTaskStatusHolder, taskState);
   }
 
-  @Timed
-  public void processStatusUpdate(Protos.TaskStatus status) {
-    long insideLock = 0;
-    final long start = schedulerLock.lock("statusUpdate");
-    try {
-      insideLock = System.currentTimeMillis();
-      unsafeProcessStatusUpdate(status);
-    } finally {
-      schedulerLock.unlock("statusUpdate", start);
-      LOG.info("Processed status update for {} ({}) in {} (waited {} for lock)", status.getTaskId().getValue(), status.getState(), JavaUtils.duration(start), JavaUtils.durationFromMillis(insideLock - start));
-    }
+  public CompletableFuture<Boolean> processStatusUpdateAsync(Protos.TaskStatus status) {
+    return statusUpdatesSemaphore.call(() -> CompletableFuture.supplyAsync(() -> {
+        long insideLock = 0;
+        final String taskId = status.getTaskId().getValue();
+        final Optional<SingularityTaskId> maybeTaskId = getTaskId(taskId);
+
+        if (!maybeTaskId.isPresent()) {
+          return false;
+        }
+
+        final long start = schedulerLock.lock(maybeTaskId.get().getRequestId(), "statusUpdate");
+        try {
```

Comment:
Does it make sense to use `callWithRequestLock()` here as well, and then make `SingularitySchedulerLock#(lock|unlock)` private?

Reply:
yes, I think I had a different exception in here at one point and didn't refactor back to callWithRequestLock, thanks for finding that